<a href="https://colab.research.google.com/github/Segtanof/pyfin/blob/main/FIN682_Exercise_6_Cross_Exchange_Arbitrage_%5BExercise%5D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# FIN 682 Exercise: Cross-Exchange Arbitrage

**Final**

*This version: 2024-04-16*

On Google Colab: [https://colab.research.google.com/drive/1Of8Jo4GG2_8ay1wNugiEXDqVXOlEUpHo](https://colab.research.google.com/drive/1Of8Jo4GG2_8ay1wNugiEXDqVXOlEUpHo)

## Why is it worth learning Python?

- *Popularity*: Python is one of the most popular programming languages ([Source](https://www.northeastern.edu/graduate/blog/most-popular-programming-languages/)) and is extensively used in the financial services industry and asset management.
- *Transferable skills*: You will acquire the concepts (such as variables, data types, loops, if-statements, ...), which other programming languages also have. Thus, if you work for a company that happens to use a different programming language, you will be able to transition quickly.
- *Package ecosystem*: Python has a rich and well-developed package ecosystem. Packages help you to use code that others have already written, which helps to save a lot of time with common problems.
- *Open source*: There is no license required to run Python. Thus, you can run it on as many computers as you like, don't need to renew it, and are not dependent on your university or employer to pay for a subscription.

### What can Python be used for in your studies?
- Empirical seminar or master theses can be done in Python.
- It could be the difference between you getting the internship or someone else.



### What can Python be used for in your future career?
- *Data analysis*: Handling, cleaning and analyzing large datasets is very easy.
- *Automation*: Due to its flexibility, almost anything can be automated in Python. For example, generating reports from a large number of input files
- *Machine learning*: Python is the most popular language for working with (and developing) machine learning models. Just ask ChatGPT "Which language was ChatGPT programmed in?" -> Python.
- *Asset management*: Many uses in asset management, such as algorithmic trading, backtesting, and portfolio optimization.

## How to get started with this exercise

### Recommended way: Google Colab

Go to the link at the top of document, add the file to your own drive ("File" -> "Save a copy to Drive") and start using it. Google's official introduction to Colab can be found here: [https://colab.research.google.com/notebook?authuser=1](https://colab.research.google.com/notebook?authuser=1)

You do **not** need to download the data file from ILIAS if you are using Google Colab.

### Alternative ways

This way is not recommended. Google Colab is significantly simpler to set up.
If you watched tutorials that do not use Colab, you should know that you can simply copy-paste the code into Colab.

If you are an advanced user, I'm sure you know how to get up and running.

## Notes
**Please note that *programming itself* is not relevant for the exam.**
You will NOT be asked something like "please write some code to do ...".
You could however be asked something about the *ideas* behind the code.

The goal of this exercise is to give interested students the possibility to work with real-world data using Python, which is widely used in the industry.

### Using Python

In general, you can find great tutorials on YouTube.
For your convenience, here are a few examples:

- [Starting point for beginners](https://www.youtube.com/watch?v=dT_XDTarsjw&t=45s) that explains data types, logic, etc.

- [Google's official introduction to Colab](https://colab.research.google.com/notebook?authuser=1) also contains some good introductory examples and explains the usage of the so-called "notebooks" and useful keyboard shortcuts

- [Overview over pandas](https://www.youtube.com/watch?v=kHHym9neKRs)
    - One of Python's main data structures are pandas DataFrames. This video gives you an overview
    - Take a look at the [official pandas cheat sheet](https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf)

If you follow a tutorial from YouTube or another source, many recommend installing VSCode, PyCharm, Anaconda, or other programs.
While these programs are great, if you are just starting out, Google Colab is the easiest choice.
**You can copy the code you have from the other tutorial into a code cell and execute it just like in the other tools.**
Take a look at Google's official introduction to Colab, which explains the concept of notebooks very well.

#### Online courses
You can also follow courses from sites such as
- [Coursera](https://www.coursera.org/courses?query=python)
- [w3school](https://www.w3resource.com/python/python-tutorial.php)
- [codecademy](https://www.codecademy.com/catalog/language/python)

<!-- ### How does programming typically work
- Use a tutorial to get up and running with the basics
- Follow the tutorial and try to *understand* why something is happening / done a certain way
- Try to change the code in the tutorials in your own notebook and check whether the outcome is as expected
- You will get a hang of the basics
- Step out of the tutorial and try different tasks, e.g. this exercise
- You WILL run into issues (sooner or later)
- Use the tools in the "Where to find more help section"
- Try to understand the suggestions you receive
- Adapt them to your code and try again -->

### Where to find more help
- For specific questions: Using your favorite search engine, "python [question]" or "pandas [question]" (if working with a pandas DataFrame) typically gives very good results. Potential sites will be:
  - StackOverflow
  - The documentation of the packages, e.g. [pandas](https://pandas.pydata.org/docs/user_guide/index.html)
- [ChatGPT](https://chat.openai.com/chat) also has the ability to produce code. Note: You should not blindly use this tool without
  - Trying yourself first (otherwise you won't think the problem through properly)
  - *Understanding* why and how the answer works
  - Keeping in mind that you will get imperfect results
  - It is very good at explaining code, so try to make use of that if you struggle with certain concepts (e.g. data types, functions, ...)
    - Example prompt: `I'm using Python. Explain how a for loop works.`
- For a broader overview: YouTube has great tutorials that cater to every taste


## General description

The raw data was downloaded from [cryptodatadownload.com](https://www.cryptodatadownload.com/data/).
We will be looking at the 2023 minute-frequency price data of Ethereum, the second largest cryptocurrency after Bitcoin.
The exchanges we study are Binance, Gemini, and Bitstamp.

If you are using Google Colab (as recommended) you do not need to use the data file. It will be downloaded automatically!

You only need to run the code below. To do so, you can use the "play"-button or you can use the keyboard shortcut "Shift+Enter".

The code blocks you see below are called "cells". When you execute a cell, only the code within the cell will be run, but no code in other cells. See [Google's official introduction to Colab](https://colab.research.google.com/notebook?authuser=1) for more details about the notebook format.

## Assumptions
We make the following assumptions. Some of them will be relaxed during the following exercises.

- Trades can be executed immediately (no delay)
- No transaction costs
- Naked short sales are possible
- You always have enough USD available on the exchanges for your purchases
- No bid-ask spread, i.e. we can trade in both directions at the given price
- No limits to arbitrage, e.g. slippage / price impact
- Trades are only executed in 2023, i.e. you cannot trade in 2022 or 2024


Each part A - E is separate from the others. The assumptions are 'reset' when moving to the next part. Within each part, assumptions are kept.

## Initialization code

In [ ]:
#@title Imports (just click the play button)

################################################################################
# Importing the modules
################################################################################
import numpy as np  # For data wrangling
import pandas as pd  # For data wrangling
import matplotlib # For plotting
import matplotlib.pyplot as plt # For plotting
import seaborn as sns  # Optional. For plotting
from scipy.optimize import bisect  # For optimization in part c

# Optional: Set some styling to make the plots look nicer
sns.set_theme(style="darkgrid")
sns.set_context("notebook")

from pathlib import Path  # For working with files and folders.

from tqdm import tqdm  # Optional: For progress bars (for intermediate to advanced users)
# Section 3.1 should be all we need for this exercise
# https://www.kaggle.com/code/nikhilkhetan/tqdm-tutorial#3.1-Iterable-based

# Set the size of plots
matplotlib.rcParams['figure.figsize'] = (10, 8)



################################################################################
# Load the input file
################################################################################


# No need to unzip the file
filename = "eth_1minute_2023.zip"

try:
    # Using Google Colab
    from google.colab import files

    # Do not change this
    datadir = Path("/content")

    fn = datadir / filename

    # Only upload the file when it doesn't exist
    if not fn.is_file():

        # Select the file to upload
        import os
        os.system('gdown 1CXtBUFGD1VpgEMAMh1Zbp8mCOygti040')


except ModuleNotFoundError:
    # Not using Google Colab
    print("It seems that you are not using Google Colab. Please use Google Colab.")

    # CHANGE THIS!
    datadir = (
        Path().cwd()
    )  # Or something like Path("c:/Program Files/PSF"). Note that backslashes "\" are entered as forward slashes "/".
    fn = datadir / filename

# Load the data into a pandas DataFrame
prices = pd.read_csv(fn, index_col=[0], parse_dates=[0])

What does the input data look like? It's a large table with 3 columns for the three exchanges.

Each row represents the price at a specific point in time in 1 minute intervals.

In [ ]:
prices.head(2)

binance   gemini  bitstamp
date                                           
2023-01-01 00:00:00  1196.13  1193.21    1195.2
2023-01-01 00:01:00  1196.09  1193.78    1195.4

## Part A (Basic strategy)

**Assumptions**

- Trades can be executed immediately (no delay)
- No transaction costs
- Naked short sales are possible
- You always have enough USD available on the exchanges for your purchases
- No bid-ask spread, i.e. we can trade in both directions at the given price
- No limits to arbitrage, e.g. slippage / price impact
- Trades are only executed in 2023, i.e. you cannot trade in 2022 or 2024

---

1. What was the highest price observed in 2023?
  *Hint: You can use the [max method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.max.html) of DataFrames*
2. What was the lowest price observed in 2023? On which exchange was it?
  *Hint: You can use the [min](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.min.html) and [idxmin](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.idxmin.html) methods of DataFrames*
3. What was the maximum price difference between the cheapest and most expensive price on each exchange? Use DataFrame methods to solve this!
4. What was the cheapest price in each minute? Show the first 10 minutes. *Hint: The min and max methods have an 'axis' keyword. You can use [iloc](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iloc.html) to get the first ten rows.*
5. What was the average price in 2023 on Gemini? *Hint: Look up the pandas documentation for something along the lines of average or mean. You can also look for the documentation on the 'loc' method.*
6. Write a function that takes an exchange as an argument and prints the average price in 2023. An example result should be "The average price on bitstamp was USD 100". Test the function with bitstamp. *Hint: Info on [functions](https://www.geeksforgeeks.org/python-functions/). Use an [f-string](https://www.geeksforgeeks.org/formatted-string-literals-f-strings-python/) for printing.*
7. Use a [for loop](https://www.geeksforgeeks.org/python-for-loops/) to run your function from part 6 for all three exchanges. *Hint: You can get the exchanges from the [column names](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.columns.html) of the DataFrame.*
8. Your trading strategy is (in each minute) to sell one unit on the most expensive exchange and at the same time buy one unit on the cheapest exchange.
  How much profit can you make over the entire year? Use the DataFrame methods from part 1 and 2. Do not use a loop.
9. Make a second DataFrame called `prices_first_10` by using the knowledge from part 4. Then, use a comparison to show which values are greater than or equal to `1196`. Store the result in `high_prices` and take a look at it. *Hint: Use a [boolean comparison](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.ge.html)*
10. Do the same for low prices below `1195`. Store it in `low_prices`.
11. Multiply the high_prices with 1 and the low_prices with -1. What do you notice compared to the orginal DataFrames from parts 9 and 10? Overwrite the DataFrames with your newly multiplied versions. *Hint: You can use the [multiply method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.multiply.html)*
12. Combine the two DataFrames by adding them. Store the result in a variable called `prices_first_10_mask`.
13. Now multiply `prices_first_10` with `prices_first_10_mask`. What do you notice? What happened to the elements?
14. A mask, called `prices_first_10_signal` is provided below. Verify that this mask has a row-wise sum of 0. Multiply this mask with `prices_first_10` and obtain the sum of the row-wise sums. What is the economic interpretation of what you calculated? How does this relate to part 8?


### 1) Highest price

### 2) Lowest price

### 3) Largest difference

### 4) Lowest price in the first 10 minutes

### 5) Average price on Gemini

### 6) Function to print average price

### 7) Loop over all exchanges

### 8) Baseline profit

### 9) Get mask of high prices


### 10) Get mask of low prices

### 11) Multiply with 1 and -1

### 12) Add the DataFrames together

### 13) Apply mask to prices_first_10

### 14) Trading signals

In [ ]:
#@title Data for part 14
# This code initializes a mask with a row-wise sum of 0.
# It represents buys (1) and sells (-1)
prices_first_10_signal = pd.DataFrame(
    {
        "binance": {
            pd.Timestamp("2023-01-01 00:00:00"): -1,
            pd.Timestamp("2023-01-01 00:01:00"): -1,
            pd.Timestamp("2023-01-01 00:02:00"): -1,
            pd.Timestamp("2023-01-01 00:03:00"): -1,
            pd.Timestamp("2023-01-01 00:04:00"): -1,
            pd.Timestamp("2023-01-01 00:05:00"): -1,
            pd.Timestamp("2023-01-01 00:06:00"): -1,
            pd.Timestamp("2023-01-01 00:07:00"): -1,
            pd.Timestamp("2023-01-01 00:08:00"): -1,
            pd.Timestamp("2023-01-01 00:09:00"): -1,
        },
        "gemini": {
            pd.Timestamp("2023-01-01 00:00:00"): 1,
            pd.Timestamp("2023-01-01 00:01:00"): 0,
            pd.Timestamp("2023-01-01 00:02:00"): 1,
            pd.Timestamp("2023-01-01 00:03:00"): 1,
            pd.Timestamp("2023-01-01 00:04:00"): 1,
            pd.Timestamp("2023-01-01 00:05:00"): 1,
            pd.Timestamp("2023-01-01 00:06:00"): 1,
            pd.Timestamp("2023-01-01 00:07:00"): 1,
            pd.Timestamp("2023-01-01 00:08:00"): 1,
            pd.Timestamp("2023-01-01 00:09:00"): 1,
        },
        "bitstamp": {
            pd.Timestamp("2023-01-01 00:00:00"): 0,
            pd.Timestamp("2023-01-01 00:01:00"): 1,
            pd.Timestamp("2023-01-01 00:02:00"): 0,
            pd.Timestamp("2023-01-01 00:03:00"): 0,
            pd.Timestamp("2023-01-01 00:04:00"): 0,
            pd.Timestamp("2023-01-01 00:05:00"): 0,
            pd.Timestamp("2023-01-01 00:06:00"): 0,
            pd.Timestamp("2023-01-01 00:07:00"): 0,
            pd.Timestamp("2023-01-01 00:08:00"): 0,
            pd.Timestamp("2023-01-01 00:09:00"): 0,
        },
    }
)


In [ ]:
# You now have the variable "prices_first_10_signal"
prices_first_10_signal.head()

## Part B (Delayed execution)

**Assumptions**

- ~~Trades can be executed immediately (no delay)~~ \[As specified\]
- No transaction costs
- Naked short sales are possible
- You always have enough USD available on the exchanges for your purchases
- No bid-ask spread, i.e. we can trade in both directions at the given price
- No limits to arbitrage, e.g. slippage / price impact
- Trades are only executed in 2023, i.e. you cannot trade in 2022 or 2024

---

A general note: Creating a DataFrame with the same rows and columns as the `prices`-DataFrame, but with buy and sell directions might be very useful.
Elementwise multiplication can then be done using the [multiply method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.multiply.html).


It might look like this (let's call it `signal` DataFrame):

| date                |  binance |     gemini |   bitstamp |
|:--------------------|---------:|-----------:|-----------:|
| 2023-01-01 00:00:00 |        1 |          0 |         -1 |
| 2023-01-01 00:01:00 |        0 |          1 |         -1 |


You should keep in mind that the signal DataFrame can represent USD or ETH trade directions. You should ensure that if you want to calculate USD profits, your signal DataFrame is set up appropriately. In that case, you should represent buying with -1 (since you have a USD outflow to pay for the Ether) and selling with 1 (since you have a USD inflow).


---

1. Now assume that your trading software takes 1 minute to execute a trade.
    This means that once you commit to your trade, you will get the next period's prices (for both buy and sell).
    Think about your expectations: What would you expect to happen to your profit when you increase the delay?
    Now calculate the profit: Is your strategy still profitable and what would the potential profit / loss be?
    Make the delay a variable in minutes.
    *Hint: The [shift method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.shift.html) might be useful*

2. Calculate the profit of your strategy for delays in the range from 1 to 60 (exclude 0) and show them in a line chart.
    Interpret your result and provide potential explanations for your observation.
    *Hint: Use a '[for loop](https://www.geeksforgeeks.org/python-for-loops/)' to go over a '[range](https://www.geeksforgeeks.org/python-range-function/)' of values; you can use [plot](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.plot.html) for the chart*

3. Your colleagues mention that as you are adding a larger delay, you are losing trades as you are shifting more and more trades into 2024, which will never be executed.
    *Explanation: With a delay of 10 minutes, you spend the first 10 minutes of 2023 doing nothing, just waiting for your software. So you essentially generate an overhang of trades that you would like to execute in 2024. By assumption, these trades will never execute -> yield a zero profit -> you lose potentially profitable trades -> with a greater delay, you mechanically see lower absolute profit.*
    
    They ask you to calculate the average profit per trade (in USD) instead and plot it dependent on the delay as in part 2. Calculate the profit for delays from 1 DAY to 60 DAYS in steps of 1 day.
    *Hint: You can use the [step size of the range function](https://www.geeksforgeeks.org/python-range-function/)*.

### 1) Profit with delay 1 minute
Now assume that your trading software takes 1 minute to execute a trade.
This means that once you commit to your trade, you will get the next period's prices.
Think about your expectations: What would you expect to happen when you increase the delay?
Now calculate the profit: Is your strategy still profitable and what would the potential profit / loss be?
Make the delay a variable in minutes.

### 2) Profit with delays in range 1 to 60 (chart)
Calculate the profit of your strategy for delays in the range from 1 to 60 (exclude 0) and show them in a line chart.
Interpret your result and provide potential explanations for your observation.

### 3) Average profit per trade by delay (chart)
Your colleagues mention that as you are adding a larger delay, you are losing observations as you are shifting more and more trades into 2024, which will never be executed.

They ask you to calculate the average profit per trade (in USD) instead and plot it dependent on the delay as in part 2. Calculate the profit for delays from 1 DAY to 60 DAYS in steps of 1 day.

## Part C (Transaction costs)

**Assumptions**

- Trades can be executed immediately (no delay)
- ~~No transaction costs~~ \[As specified\]
- Naked short sales are possible
- You always have enough USD available on the exchanges for your purchases
- No bid-ask spread, i.e. we can trade in both directions at the given price
- No limits to arbitrage, e.g. slippage / price impact
- Trades are only executed in 2023, i.e. you cannot trade in 2022 or 2024

---


1. After trading for a while, you realize that transaction costs are eating into your profits.
    Assume again that your software does not incur any delay and that you are trading exactly as before, i.e. ignore any transaction costs in your *decision making*.
    With transaction costs of [0.4%](https://www.bitstamp.net/fee-schedule/) for each transaction (buy and sell), how much of your profit is left?
    *Hint: You can directly incorporate the transaction costs in the 'signal' DataFrame that we created in part B.*

2. What is the break even amount of transaction costs (in %) before your trading strategy becomes unprofitable?
    *Hint: You can use `from scipy.optimize import bisect`. See the [manual](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.bisect.html) for details. Important: You need to give it a __function__ to optimize*
    
3. You think that you will be better off if you take transaction costs into account when making a trading decision.
    Now, you only trade when your ex-ante expected profit (with transaction costs as of part 1) is positive.
    How much profit do you make?

4. So far, we have ignored the bid-ask spread.
    Briefly describe how a non-zero bid-ask spread could affect your results.
    Also discuss slippage and the fundamental difference to transaction costs and the bid-ask spread.

In [ ]:
transaction_costs = 0.4  # = 0.4%

### 1) Basic transaction costs
After trading for a while, you realize that transaction costs are eating into your profits.
Assume again that your software does not incur any delay and that you are trading exactly as before, i.e. ignore any transaction costs in your *decision making*.
With transaction costs of [0.4%](https://www.bitstamp.net/fee-schedule/) for each transaction (buy and sell), how much of your profit is left?

### 2) Break even transaction costs
What is the break even amount of transaction costs (in %) before your trading strategy becomes unprofitable?

### 3) Ex-ante positive profit trades only
You think that you will be better off if you take transaction costs into account when making a trading decision.
Now, you only trade when your ex-ante expected profit (with transaction costs as of part 1) is positive.
How much profit do you make?

### 4) Bid-ask spread and slippage
So far, we have ignored the bid-ask spread.
Briefly describe how a non-zero bid-ask spread could affect your results.
Also discuss slippage and the fundamental difference to transaction costs and the bid-ask spread.

## Part D (Trade balance visualizations)

**Assumptions**

- Trades can be executed immediately (no delay)
- No transaction costs
- Naked short sales are possible
- You always have enough USD available on the exchanges for your purchases
- No bid-ask spread, i.e. we can trade in both directions at the given price
- No limits to arbitrage, e.g. slippage / price impact
- Trades are only executed in 2023, i.e. you cannot trade in 2022 or 2024

---


Background knowledge:
Transferring ETH between exchanges is far from instant.
When sending funds, the current overall Ethereum network load and the transaction costs you are willing to pay determine how fast your transaction goes through.
Because the receiving exchange will only allow you to trade with the received ETH once a certain number of "confirmations" are received, you will have to wait until several Ethereum miners confirm the transaction.
According to [this source](https://blog.tezro.com/cryptocurrency-transaction-speeds/), in 2023, the average confirmation time for Ethereum was 6 minutes, and thus far away from instant.

To circumvent this bottleneck, you have a few ideas:


1. You decide to keep a balance of the cryptocurrency on each exchange (in addition to the (assumed) USD you keep on the exchange).
    Which potential types of risk are you now exposed to?
    How do (or would) funds handle these risks on stock exchanges?
    Why or why not could this still be considered arbitrage in the classic sense?

2. To get a feeling for this risk, you want to visualize the balance of your account (**in ETH**) after 14 days of trading at `2023-01-15 00:00:00` given an initial investment of USD 1,000,000 (that you use to buy ETH with), which is split equally across all exchanges (\$333,333 deposited on each exchange).
    No rebalancing is being done.
    Show the information in a bar chart. The exchanges should be on the x-axis, the current balance on the y-axis. Add a horizontal line denoting the initial balance.
    What do you notice? Interpret your results.
    *Hint: The [cumsum method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.cumsum.html) might be useful*.
    
3. Your colleagues mention that the time dimension is missing.
    They think a line chart showing the balance on each of the exchanges over the full year would give them a better overview.

### 1) Potential risks
You decide to keep a balance of the cryptocurrency on each exchange (in addition to the USD you keep on the exchange).
Which potential types of risk are you now exposed to?
How do (or would) funds handle these risks on stock exchanges?
Why or why not could this still be considered arbitrage in the classic sense?

### 2) Balance snapshot at a specific point in time

To get a feeling for this risk, you want to visualize the balance of your account (in ETH) after 14 days of trading at `2023-01-15 00:00:00` given an initial investment of USD 1,000,000 (that you use to buy ETH with), which is split equally across all exchanges (\$333,333 deposited on each exchange).
No rebalancing is being done.
Show the information in a bar chart. The exchanges should be on the x-axis, the current balance on the y-axis. Add a horizontal line denoting the initial balance.
What do you notice? Interpret your results.

In [ ]:
initial_investment = 1_000_000

### 3) Balance over time
Your colleagues mention that the time dimension is missing.
They think a line chart showing the balance on each of the exchanges over the full year would give them a better overview.

## Part E (Redistribution)

**Assumptions**

- Trades can be executed immediately (no delay)
- No transaction costs **for trades. Withdrawal fees and transfer costs apply as specified in the task**
- ~~Naked short sales are possible~~
- You always have enough USD available on the exchanges for your purchases
- No bid-ask spread, i.e. we can trade in both directions at the given price
- No limits to arbitrage, e.g. slippage / price impact
- Trades are only executed in 2023, i.e. you cannot trade in 2022 or 2024

---

We assume that you cannot sell crypto assets if you have a zero balance on an exchange (i.e. no naked short sales).
Thus, you need to redistribute periodically in order to have enough ETH on each exchange.
    
    
You decide to redistribute whenever your ETH balance on an exchange falls **below** 50 ETH.
As in part d, you initially purchase ETH worth USD 1,000,000 split across the three exchanges.
Costs are expected to be 0.01 ETH for the withdrawal and the transfer fees (per-transaction, i.e. independent of value).
No fees are expected for deposits.
Redistributed funds are available in the next period.

You redistribute in such a way that you have the same number of ETH on each exchange (before transfer costs).
This is to say that if you hold 0 ETH on A, 1 ETH on B and 2 ETH on C, you would want a 1 / 1 / 1 split after redistributing.
For simplicitly, we say that we will transfer 1 ETH from C to A, paying 0.01 ETH in fees on C, such that we end up with 1 / 1 / 0.99 (If you find it easier to implement the transfer fees to be paid on the receiving exchange, feel free to do it that way). Assume that you always need 2 transactions per redistribution (i.e. no exchange is perfectly balanced).
    
1. In the worst case, how many minutes could you still trade before you run out of ETH to sell after the balance falls below the rebalancing threshold? No code required for this part.

2. How often do you redistribute (number of redistributions in 2023) and which costs do you incur in USD (value of the crypto assets at the initiation time of the rebalancing transaction as priced at the sending exchange)?
    *Hint: If you solve this with a loop, while you are trying to find the solution, limit the data to e.g. January to ensure faster computation.*

### 1) Time left
In the worst case, how many minutes could you still trade before you run out of ETH to sell after the balance falls below the rebalancing threshold?

### 2) Number and cost of redistributions
How often do you redistribute (number of redistributions in 2023) and which costs do you incur in USD (value of the crypto assets at the initiation time of the rebalancing transaction as priced at the sending exchange)?

In [ ]:
rebalancing_threshold = 50  # in ETH
withdrawal_fee = 0.01